src/pkgcheck/checks/ruby.py

```diff
@@ -0,0 +1,107 @@
+import itertools
+
+from pkgcore.ebuild.atom import atom
+from snakeoil.sequences import iflatten_instance
+from snakeoil.strings import pluralism
+
+from .. import results
+from . import Check
+
+
+IUSE_PREFIX = "ruby_targets_"
+
+
+class RubyCompatUpdate(results.VersionResult, results.Info):
+    """``USE_RUBY`` can be updated to support newer python version(s)."""
+
+    def __init__(self, updates, **kwargs):
+        super().__init__(**kwargs)
+        self.updates = tuple(updates)
+
+    @property
+    def desc(self):
+        s = pluralism(self.updates)
+        updates = ", ".join(self.updates)
+        return f"USE_RUBY update{s} available: {updates}"
+
+
+class RubyCompatCheck(Check):
+    """Check ruby ebuilds for possible ``USE_RUBY`` updates.
+
+    Supports ebuilds inheriting ``ruby-ng``.
+    """
+
+    known_results = frozenset({RubyCompatUpdate})
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        repo = self.options.target_repo
+        # sorter for ruby targets leveraging USE_EXPAND flag ordering from repo
+        self.sorter = repo.use_expand_sorter("ruby_targets")
+
+        # determine available USE_RUBY use flags
+        targets = []
+        for target, _desc in repo.use_expand_desc.get(IUSE_PREFIX[:-1], ()):
+            if target[len(IUSE_PREFIX) :].startswith("ruby"):
+                targets.append(target[len(IUSE_PREFIX) :])
+        self.multi_targets = tuple(sorted(targets, key=self.sorter))
+
+    def ruby_deps(self, deps, prefix):
+        for dep in (x for x in deps if x.use):
+            for x in dep.use:
+                if x.startswith(("-", "!")):
+                    continue
+                if x.startswith(prefix):
+                    yield dep.no_usedeps
+                    break
+
+    def deps(self, pkg):
+        """Set of dependencies for a given package's attributes."""
+        return {
+            p
+            for attr in (x.lower() for x in pkg.eapi.dep_keys)
+            for p in iflatten_instance(getattr(pkg, attr), atom)
+            if not p.blocks
+        }
+
+    def feed(self, pkg):
+        if "ruby-ng" not in pkg.inherited:
+            return
+
+        deps = self.deps(pkg)
+
+        try:
+            # determine the latest supported python version
+            latest_target = sorted(
+                (
+                    f"ruby{x.slot.replace('.', '')}"
+                    for x in deps
+                    if x.key == "dev-lang/ruby" and x.slot is not None
```

Comment:
we should probably ignore virtual/rubygems and virtual/ruby-ssl too

Reply:
done